### PR TITLE
Laboratory Work #1, Vita Zaytseva - 20FPL2

### PR DESCRIPTION
**Pull request by vitazaytseva**

### CODE REVIEW

File: seminars/03.04.2022/index.html [hunk 1]
```diff
@@ -0,0 +1,316 @@
+<!DOCTYPE html>
```
Comment (demid5111) [inline seminars/03.04.2022/index.html:1]: you cannot commit unrelated files. discard

File: seminars/03.04.2022/index.html [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+<html lang="ru" dir="ltr" prefix="content: http://purl.org/rss/1.0/modules/content/  dc: http://purl.org/dc/terms/  foaf: http://xmlns.com/foaf/0.1/  og: http://ogp.me/ns#  rdfs: http://www.w3.org/2000/01/rdf-schema#  schema: http://schema.org/  sioc: http://rdfs.org/sioc/ns#  sioct: http://rdfs.org/sioc/types#  skos: http://www.w3.org/2004/02/skos/core#  xsd: http://www.w3.org/2001/XMLSchema# " class="no-js adaptivetheme">
+  <head>
```
Comment (demid5111) [inline seminars/03.04.2022/index.html:3]: the same for logo

File: seminars/03.04.2022/try_requests.py [hunk 3]
```diff
@@ -4,71 +4,87 @@
 import requests
 from requests import Timeout, HTTPError
 
-if __name__ == '__main__':
```
Comment (demid5111) [inline seminars/03.04.2022/try_requests.py:7]: you cannot commit changes to the seminars listings. instead move the code to the place it is expected


carefully read the docs. first part of assignment should be made within `scrapper.py` module.

File: seminars/03.04.2022/try_requests.py [hunk 4]
```diff
@@ -4,71 +4,87 @@
...
+        f.write(response.content)
 
-    # 3.3 saving binary files formats, such as images
-    response = requests.get('https://pypi.org/static/images/logo-small.95de8436.svg')
 
-    with open('logo.svg', 'wb') as f:
-        f.write(response.content)
+main()
+
+# if __name__ == '__main__':
```
Comment (demid5111) [inline seminars/03.04.2022/try_requests.py:23]: especially we never commit commented code, maybe rarely and only by mistake. please discard

**Review by demid5111 — CHANGES_REQUESTED** [review]
great start!

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+requests
```
Comment (demid5111) [inline requirements.txt:1]: better to specify particular versions

File: scrapper.py [hunk 6]
```diff
@@ -58,6 +59,20 @@ def validate_config(crawler_path):
...
+    with open('index.html', 'w', encoding='utf-8') as f:
+        f.write(response.text)
+    response = requests.get('https://russkayarech.ru/images/topsite2.png')
+    with open('logo.png', 'wb') as f:
+        f.write(response.content)
+
+
 if __name__ == '__main__':
     # YOUR CODE HERE
     pass
```
Comment (demid5111) [inline scrapper.py:77]: remove unnecessary stubs

File: scrapper.py [hunk 7]
```diff
@@ -58,6 +59,20 @@ def validate_config(crawler_path):
...
 
 
+def main():
+    response = requests.get('https://russkayarech.ru/ru')
```
Comment (demid5111) [inline scrapper.py:63]: I see that you are able to download a page. But this is far not enough to hand over programming assignment

**Review by demid5111 — CHANGES_REQUESTED** [review]
it is time to start. there is plenty of things to do

**Comment by demid5111** [comment]
you are at risk of missing a deadline! hurry up!

File: scrapper.py [hunk 8]
```diff
@@ -25,17 +39,37 @@ class Crawler:
...
 
     def _extract_url(self, article_bs):
-        pass
+        urls_bs = article_bs.find_all('td', class_='views-field views-field-title table__cell')
+        full_urls = []
+        for url_bs in urls_bs:
+            url = url_bs.find('a')['href']
+            full_urls.append(HTTP_PATTERN + url)
+        for full_url in full_urls:
+            if len(self.urls) >= 100:
```
Comment (demid5111) [inline scrapper.py:55]: number of articles to collect should be controlled by a configuration file

File: scrapper.py [hunk 9]
```diff
@@ -44,20 +78,92 @@ def get_search_urls(self):
...
+        self.article_id = article_id
+        self.article = Article(self.article_url, self.article_id)
+
+    def _fill_article_with_text(self, article_bs):
+        pdf_url = article_bs.find('iframe', class_='pdf')['data-src']
+        pdf = PDFRawFile(pdf_url, self.article_id)
+        pdf.download()
+
+        pdf_with_lit = pdf.get_text()
+        pdf_without_lit = pdf_with_lit.split('Литература')[0]
```
Comment (demid5111) [inline scrapper.py:93]: why do we need an intermediate variable?

File: scrapper.py [hunk 10]
```diff
@@ -44,20 +78,92 @@ def get_search_urls(self):
...
+        self._fill_article_with_meta_information(article_bs)
+        return self.article
+
+
 def prepare_environment(base_path):
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    path_to_base_path = Path(base_path)
```
Comment (demid5111) [inline scrapper.py:122]: are there cases when `base_path` is not a Path instance?

File: scrapper.py [hunk 11]
```diff
@@ -44,20 +78,92 @@ def get_search_urls(self):
...
+    with open(crawler_path) as file:
+        configuration = json.load(file)
+
+    http_pattern = re.compile(HTTP_PATTERN)
+    for url in configuration["seed_urls"]:
+        result = http_pattern.search(url)
+        if not result:
+            raise IncorrectURLError
+
+    seed_urls = configuration["seed_urls"]
```
Comment (demid5111) [inline scrapper.py:141]: is it guaranteed to have `seed_urls `and `total_articles_to_find_and_parse` in a configuration file?

File: scrapper.py [hunk 12]
```diff
@@ -44,20 +78,92 @@ def get_search_urls(self):
...
-    # YOUR CODE HERE
-    pass
+    main_seed_urls, main_max_articles = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+
+    crawler = Crawler(main_seed_urls, main_max_articles)
+    crawler.find_articles()
+
+    ID_OF_ARTICLE = 1
+    for main_article_url in crawler.urls:
```
Comment (demid5111) [inline scrapper.py:164]: is there a way to get the index of the current element without creating a counter outside the loop?

**Review by demid5111 — CHANGES_REQUESTED** [review]
I have not seen the tests being run. Have you looked into the problem or asked at the forum? without tests results you will not be accepted for oral hand over

File: scrapper.py [hunk 13]
```diff
@@ -44,20 +72,95 @@ def get_search_urls(self):
         pass
```
Comment (demid5111) [inline scrapper.py:72]: it is not implemented

File: scrapper.py [hunk 14]
```diff
@@ -44,20 +72,95 @@ def get_search_urls(self):
...
-    pass
+    main_seed_urls, main_max_articles = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+
+    crawler = Crawler(main_seed_urls, main_max_articles)
+    crawler.find_articles()
+
+    for id_of_article, url_of_article in enumerate(crawler.urls):
+        article_parser = HTMLParser(article_url=url_of_article, article_id=id_of_article + 1)
+        article = article_parser.parse()
```
Comment (demid5111) [inline scrapper.py:163]: parse method also makes requests to the website, we need to make a request human-like as well (as you did in crawler)

File: scrapper.py [hunk 15]
```diff
@@ -44,20 +72,95 @@ def get_search_urls(self):
...
+
+    def _fill_article_with_meta_information(self, article_bs):
+        self.article.author = article_bs.find('span', class_='field__item-wrapper').text
+
+        self.article.title = article_bs.find('span', class_='field field-name-title field-formatter-string '
+                                                            'field-type-string field-label-hidden').text
+
+        node_content = article_bs.find('div', class_='node__content clearfix')
+        year = node_content.find('div').find_next_siblings()[2].text.strip('\n').strip(' ')[:4]
+        self.article.date = datetime.date(int(year), 1, 1)
```
Comment (demid5111) [inline scrapper.py:96]: month should be retrieved from an issue number

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice progress!

File: scrapper.py [hunk 16]
```diff
@@ -44,20 +72,95 @@ def get_search_urls(self):
...
+    http_pattern = re.compile(HTTP_PATTERN)
+    for url in configuration["seed_urls"]:
+        result = http_pattern.search(url)
+        if not result:
+            raise IncorrectURLError
+
+    seed_urls = configuration["seed_urls"]
+    total_articles_to_find_and_parse = configuration["total_articles_to_find_and_parse"]
+
+    if 'seed_urls' not in configuration:
```
[the remark this answers is not in the document]
Comment (vitazaytseva) [reply scrapper.py:136]: Because here before checking for the existence of a key, we check for seed urls under that key. I've realized my mistake, redo

**Review by demid5111 — CHANGES_REQUESTED** [review]
task: print length of literature, number of pages of PDF file

File: scrapper.py (hunk 13 above)
Comment (demid5111) [reply scrapper.py:71]: still applies

**Review by demid5111 — CHANGES_REQUESTED** [review]
still several comments about scrapper and no progress with a pipeline!

File: config/target_score.txt [hunk 17]
```diff
@@ -1,5 +1,5 @@
 # Target score for scrapper.py:
-4
+8
 
 # Target score for pipeline.py:
-0
+6
```
Comment (demid5111) [inline config/target_score.txt]: you should target 10! this is absolutely doable for a student like you

File: pipeline.py [hunk 18]
```diff
@@ -51,19 +60,26 @@ class CorpusManager:
...
+        self._storage = {}
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        path = Path(self.path_to_raw_txt_data)
+        for file in path.glob("*.txt"):
+            pattern = re.search(r'\d+', file.name)
```
Comment (demid5111) [inline pipeline.py]: exactly what I was talking about today: pattern in a loop

File: pipeline.py [hunk 19]
```diff
@@ -51,19 +60,26 @@ class CorpusManager:
...
+            if pattern:
```
Comment (demid5111) [inline pipeline.py]: make code less nested

File: pipeline.py [hunk 20]
```diff
@@ -72,31 +88,82 @@ class TextProcessingPipeline:
...
+            raw_text = article.get_raw_text()
+            tokens = self._process(raw_text)
+            cleaned_tokens = []
+            single_tagged_tokens = []
+            for token in tokens:
+                cleaned_tokens.append(token.get_cleaned())
+                single_tagged_tokens.append(token.get_single_tagged())
+            article.save_as(*cleaned_tokens, kind=ArtifactType.cleaned)
+            article.save_as(*single_tagged_tokens, kind=ArtifactType.single_tagged)
+            print(*cleaned_tokens)
```
Comment (demid5111) [inline pipeline.py]: rm

File: pipeline.py [hunk 21]
```diff
@@ -72,31 +88,82 @@ class TextProcessingPipeline:
...
+        articles = self.corpus_manager.get_articles().values()
+        for article in articles:
+            raw_text = article.get_raw_text()
+            tokens = self._process(raw_text)
+            cleaned_tokens = []
+            single_tagged_tokens = []
+            for token in tokens:
+                cleaned_tokens.append(token.get_cleaned())
+                single_tagged_tokens.append(token.get_single_tagged())
+            article.save_as(*cleaned_tokens, kind=ArtifactType.cleaned)
```
Comment (demid5111) [inline pipeline.py]: why do you unpack list here?

Comment (demid5111) [reply pipeline.py]: save_as accepts a string, are you passing here a string?

File: pipeline.py [hunk 22]
```diff
@@ -72,31 +88,82 @@ class TextProcessingPipeline:
...
-    pass
+    path = Path(path_to_validate)
+    if not path.exists():
+        raise FileNotFoundError
+    if not path.is_dir():
+        raise NotADirectoryError
+
+    all_article_ids = []
+    for file in path.glob("*.txt"):
+        name_pattern = re.match(r'\d+_(raw.txt)', file.name)
```
Comment (demid5111) [inline pipeline.py]: why do you make a group from raw.txt?

File: requirements.txt [hunk 23]
```diff
@@ -1 +1,7 @@
-
+beautifulsoup4==4.10.0
+lxml==4.8.0
+PyMuPDF==1.19.6
+requests==2.27.1
+wget==3.2
+pymystem3==0.2.0
```
Comment (demid5111) [inline requirements.txt]: order?

File: pipeline.py (hunk 22 above)
Comment (demid5111) [inline pipeline.py]: you could already extract ID instead of creating several patterns, and even worse - in a loop

File: pipeline.py [hunk 24]
```diff
@@ -72,31 +88,82 @@ class TextProcessingPipeline:
...
         Processes each token and creates MorphToken class instance
         """
-        pass
+        mystem = Mystem()
+        analyzed_text = mystem.analyze(raw_text)
+        tokens = []
+        for analyzed_word in analyzed_text:
+            if not analyzed_word.get('analysis'):
+                continue
+            morphological_token = MorphologicalToken(original_word=analyzed_word.get('text'))
```
Comment (demid5111) [inline pipeline.py]: all pieces are very fragile - you might have empty dictionaries and what happens then?

Comment (demid5111) [reply pipeline.py]: or analysis lists might be empty

**Review by demid5111 — CHANGES_REQUESTED** [review]
still expecting implementation for a higher score. I believe in you

File: pipeline.py [hunk 25]
```diff
@@ -1,6 +1,12 @@
 """
 Pipeline for text processing implementation
 """
+from pathlib import Path
```
Comment (demid5111) [inline pipeline.py]: and, finally, order of imports

**Review by demid5111 — CHANGES_REQUESTED** [review]
keep going, only one day is left. make sure you address all the comments and pipeline is green by Friday

**Pull request closed without merging**